File: jetbrains-plugin/src/main/kotlin/com/mariadbprofiler/plugin/service/LogParserService.kt
package com.mariadbprofiler.plugin.service

import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.mariadbprofiler.plugin.model.QueryEntry
import kotlinx.serialization.json.Json
import java.io.File
import java.io.RandomAccessFile
import java.nio.charset.StandardCharsets

@Service(Service.Level.PROJECT)
class LogParserService(private val project: Project) {

    private val log = Logger.getInstance(LogParserService::class.java)

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    fun parseJsonlFile(filePath: String): List<QueryEntry> {
        val file = File(filePath)
        if (!file.exists() || !file.canRead()) {
            return emptyList()
        }

        val entries = mutableListOf<QueryEntry>()
        var parseErrors = 0
        file.useLines { lines ->
            lines.forEachIndexed { index, line ->
                val trimmed = line.trim()
                if (trimmed.isNotEmpty()) {
                    try {
                        entries.add(json.decodeFromString<QueryEntry>(trimmed))
                    } catch (e: Exception) {
                        log.debug("Failed to parse line $index in $filePath: ${e.message}")
                        parseErrors++
                    }
                }
            }
        }
        if (parseErrors > 0) {
            val errorLog = project.getService(ErrorLogService::class.java)
            errorLog.addWarning("LogParser", "$parseErrors line(s) failed to parse in ${file.name}")
        }
        return entries
    }

    /**
     * Read JSONL entries starting from byte offset.
     * Returns (entries, newOffset) where newOffset is the file size at read time.
     */
    fun parseJsonlFileFromOffset(filePath: String, offset: Long): Pair<List<QueryEntry>, Long> {
        val file = File(filePath)
        if (!file.exists() || !file.canRead()) {
            return Pair(emptyList(), offset)
        }

        val fileSize = file.length()
        if (fileSize <= offset) {
            return Pair(emptyList(), offset)
        }

        val entries = mutableListOf<QueryEntry>()
        RandomAccessFile(file, "r").use { raf ->
            raf.seek(offset)
            val bytes = ByteArray((fileSize - offset).toInt())
            raf.readFully(bytes)
            val content = String(bytes, StandardCharsets.UTF_8)
            for (line in content.lines()) {
                val trimmed = line.trim()
                if (trimmed.isNotEmpty()) {
                    try {
                        entries.add(json.decodeFromString<QueryEntry>(trimmed))
                    } catch (e: Exception) {
                        log.debug("Failed to parse incremental line: ${e.message}")
                    }
                }
            }
        }
        return Pair(entries, fileSize)
    }

    fun readRawLog(filePath: String, maxLines: Int = 500): List<String> {
        val file = File(filePath)
        if (!file.exists() || !file.canRead()) {
            return emptyList()
        }

        val buffer = ArrayDeque<String>(maxLines)
        file.useLines { lines ->
            lines.forEach { line ->
                if (buffer.size >= maxLines) {
                    buffer.removeFirst()
                }
                buffer.addLast(line)
            }
        }
        return buffer.toList()
    }

    /**
     * Read raw log lines starting from byte offset.
     * Returns (lines, newOffset) where newOffset is the file size at read time.
     */
    fun tailRawLog(filePath: String, fromOffset: Long): Pair<List<String>, Long> {
        val file = File(filePath)
        if (!file.exists() || !file.canRead()) {
            return Pair(emptyList(), fromOffset)
        }

        val fileSize = file.length()
        if (fileSize <= fromOffset) {
            return Pair(emptyList(), fromOffset)
        }

        val lines = mutableListOf<String>()
        RandomAccessFile(file, "r").use { raf ->
            raf.seek(fromOffset)
            val bytes = ByteArray((fileSize - fromOffset).toInt())
            raf.readFully(bytes)
            val content = String(bytes, StandardCharsets.UTF_8)
            for (line in content.lines()) {
                if (line.isNotEmpty()) {
                    lines.add(line)
                }
            }
        }
        return Pair(lines, fileSize)
    }
}

File: vscode-extension/src/service/LogParserService.ts
import * as fs from 'fs';
import * as vscode from 'vscode';
import { QueryEntry, RawQueryEntry, fromRaw } from '../model/QueryEntry';

export class LogParserService {
  private errorChannel: vscode.OutputChannel;

  constructor(errorChannel: vscode.OutputChannel) {
    this.errorChannel = errorChannel;
  }

  parseJsonlFile(filePath: string): QueryEntry[] {
    if (!fs.existsSync(filePath)) { return []; }

    const content = fs.readFileSync(filePath, 'utf-8');
    return this.parseJsonlContent(content);
  }

  parseJsonlFileFromOffset(filePath: string, offset: number): { entries: QueryEntry[]; newOffset: number } {
    if (!fs.existsSync(filePath)) {
      return { entries: [], newOffset: offset };
    }

    const stat = fs.statSync(filePath);
    if (stat.size <= offset) {
      return { entries: [], newOffset: offset };
    }

    const fd = fs.openSync(filePath, 'r');
    try {
      const bufSize = stat.size - offset;
      const buffer = Buffer.alloc(bufSize);
      fs.readSync(fd, buffer, 0, bufSize, offset);
      const content = buffer.toString('utf-8');
      const entries = this.parseJsonlContent(content);
      return { entries, newOffset: stat.size };
    } finally {
      fs.closeSync(fd);
    }
  }

  readRawLogTail(filePath: string, maxLines: number = 500): string {
    if (!fs.existsSync(filePath)) { return ''; }

    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');
    const tail = lines.slice(-maxLines);
    return tail.join('\n');
  }

  tailRawLog(filePath: string, offset: number): { content: string; newOffset: number } {
    if (!fs.existsSync(filePath)) {
      return { content: '', newOffset: offset };
    }

    const stat = fs.statSync(filePath);
    if (stat.size <= offset) {
      return { content: '', newOffset: offset };
    }

    const fd = fs.openSync(filePath, 'r');
    try {
      const bufSize = stat.size - offset;
      const buffer = Buffer.alloc(bufSize);
      fs.readSync(fd, buffer, 0, bufSize, offset);
      return { content: buffer.toString('utf-8'), newOffset: stat.size };
    } finally {
      fs.closeSync(fd);
    }
  }

  private parseJsonlContent(content: string): QueryEntry[] {
    const entries: QueryEntry[] = [];
    const lines = content.split('\n');

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) { continue; }

      try {
        const raw: RawQueryEntry = JSON.parse(trimmed);
        entries.push(fromRaw(raw));
      } catch (e) {
        this.errorChannel.appendLine(`[LogParser] Failed to parse line: ${trimmed.substring(0, 100)}`);
      }
    }

    return entries;
  }
}
